Refactor StableMIR

makai410 opened the pull request:
This PR refactors stable-mir according to the guidance in [this doc](https://hackmd.io/jBRkZLqAQL2EVgwIIeNMHg). It reverses the dependency between `rustc_smir` and `stable_mir`, making `rustc_smir` completely agnostic  of `stable_mir`.

Under the new architecture, the `rustc_smir` crate would retain direct access to rustc queries, while `stable_mir` should proxy all such requests through `rustc_smir` instead of accessing rustc's internals directly. `stable_mir` would only be responsible for the conversion between internal and stable constructs.

This PR mainly introduces these changes:

- **Bridge / Tables<'tcx, B: Bridge>**

```rust
/// A trait defining types that are used to emulate StableMIR components, which is really
/// useful when programming in stable_mir-agnostic settings.
pub trait Bridge {
    type DefId: Copy + Debug + PartialEq + IndexedVal;
    type AllocId: Copy + Debug + PartialEq + IndexedVal;
    type Span: Copy + Debug + PartialEq + IndexedVal;
    type Ty: Copy + Debug + PartialEq + IndexedVal;
    type InstanceDef: Copy + Debug + PartialEq + IndexedVal;
    type TyConstId: Copy + Debug + PartialEq + IndexedVal;
    type MirConstId: Copy + Debug + PartialEq + IndexedVal;
    type Layout: Copy + Debug + PartialEq + IndexedVal;
    type Error: SmirError;
}

pub struct Tables<'tcx, B: Bridge> {
    tcx: TyCtxt<'tcx>,
    pub(crate) def_ids: IndexMap<DefId, B::DefId>,
    pub(crate) alloc_ids: IndexMap<AllocId, B::AllocId>,
    pub(crate) spans: IndexMap<rustc_span::Span, B::Span>,
    pub(crate) types: IndexMap<Ty<'tcx>, B::Ty>,
    pub(crate) instances: IndexMap<ty::Instance<'tcx>, B::InstanceDef>,
    pub(crate) ty_consts: IndexMap<ty::Const<'tcx>, B::TyConstId>,
    pub(crate) mir_consts: IndexMap<mir::Const<'tcx>, B::MirConstId>,
    pub(crate) layouts: IndexMap<rustc_abi::Layout<'tcx>, B::Layout>,
}
```

Since `rustc_smir` needs these stable types somewhere, using associated types is a good approach.

- **SmirContainer / SmirInterface**

```rust
/// A container which is used for TLS.
pub struct SmirContainer<'tcx, B: Bridge> {
    pub tables: RefCell<Tables<'tcx, B>>,
    pub cx: RefCell<SmirCtxt<'tcx, B>>,
}

impl<'tcx> SmirInterface for SmirContainer<'tcx, BridgeTys> {
    // ...
}

/// Provides direct access to rustc's internal queries.
///
/// The [`crate::stable_mir::compiler_interface::SmirInterface`] must go through
/// this context to obtain rustc-level information.
pub struct SmirCtxt<'tcx, B: Bridge> {
    tcx: TyCtxt<'tcx>,
    _marker: PhantomData<B>,
}
```

This PR moves `Tables` from `SmirCtxt` to a new `SmirContainer` struct, since mutable borrows of `tables` should only be managed by `SmirInterface`. This change prevents `SmirCtxt` from holding separate borrows and requires passing `tables` explicitly when needed:
```rust
impl<'tcx, B: Bridge> SmirCtxt<'tcx, B> {
    // ...
    /// Get the body of an Instance which is already monomorphized.
    pub fn instance_body(
        &self,
        instance: ty::Instance<'tcx>,
        tables: &mut Tables<'tcx, B>,
    ) -> Option<Body<'tcx>> {
        tables
            .instance_has_body(instance)
            .then(|| BodyBuilder::new(self.tcx, instance).build(tables))
    }
    // ...
}
```



This PR introduces `SmirContainer` as a separate struct rather than bundling it into a `SmirInterface` struct. This separation makes the architecture more modular and easier to reason about.

- **context/traits.rs**

We use this file to define traits that are used for encapsulating the associated functions in the rustc's internals. This is much easier to use and maintain than directly cramming everything into `SmirCtxt`. Here is a real-world use case:
```rust
impl RustcInternal for ExistentialTraitRef {
    type T<'tcx> = rustc_ty::ExistentialTraitRef<'tcx>;

    fn internal<'tcx>(
        &self,
        tables: &mut Tables<'_, BridgeTys>,
        cx: &SmirCtxt<'tcx, BridgeTys>,
    ) -> Self::T<'tcx> {
        use rustc_smir::context::SmirExistentialTraitRef;
        cx.new_from_args(self.def_id.0.internal(tables, cx), self.generic_args.internal(tables, cx))
    }
}
```

- **Separation of `rustc_smir::alloc`**

The previous `rustc_smir::alloc` had many direct calls to rustc queries. This PR splits it into two parts: `rustc_smir::alloc` and `stable_mir::alloc`. Following the same pattern as `SmirCtxt` and `SmirInterface`, the `rustc_smir::alloc` handles all direct interactions with rustc queries and performs the actual memory allocations, while the `stable_mir::alloc` is responsible for constructing stable components.

- **Removal of `convert/error.rs`**

We use `SmirError::from_internal` instead, since implementing `Stable` for these internal errors would be redundant—`tables` is not actually used. If we later need to add something like `LayoutError` to `stable_mir`, we could implement it as follows:
```rust
impl SmirError for stable_mir::LayoutError {
    fn from_internal<T: Debug>(err: T) -> Self {
        // ...
    }
}
```



**Unresolved questions:**

- There are still a few direct calls to rustc's internals scattered across `impl Stable`s, but most of them appear to be relatively stable, e.g., `mir::interpret::ConstAllocation::inner(self)` and `mir::syntax::SwitchTargets::otherwise(self)`.

r? @celinval

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14823922969/job/41614584787) [(plain)](https://github.com/rust-lang/rust/commit/28f5ec1f45171820daec7b70c669e68f25039d09/checks/41614584787/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting rustc_hir_pretty v0.0.0 (/checkout/compiler/rustc_hir_pretty)
error: unresolved link to `SmirCtxt`
   --> compiler/rustc_smir/src/stable_mir/compiler_interface.rs:171:48
    |
171 | /// All queries are delegated to an internal [`SmirCtxt`] that provides
    |                                                ^^^^^^^^ no item named `SmirCtxt` in scope
    |
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: could not document `rustc_smir`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:32
  local time: Sun May  4 18:41:39 UTC 2025
  network time: Sun, 04 May 2025 18:41:39 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

makai410 committed:
refactor: add `Tables<'tcx, B: Bridge>` and `SmirContainer`

makai410 committed:
refactor: make `SmirInterface` a trait and impl it for `SmirContainer`

- rewrite all `SmirInterface` apis.
- add `BridgeTys` to impl those associated types in `Bridge`.
- move `**_def()` stuffs living in `impl Tables` from `rustc_internal` to `stable_mir`.

makai410 committed:
refactor: move `IndexedVal` from `stable_mir` to `rustc_smir`

makai410 committed:
refactor: impl `SmirError` for `stable_mir::Error`

makai410 committed:
refactor: split `rustc_smir::alloc` into two parts

The previous `rustc_smir::alloc` had many direct calls to rustc queries.

This commit splits it into two parts: `rustc_smir::alloc` and `stable_mir::alloc`.

Following the same pattern as `SmirCtxt` and `SmirInterface`, the `rustc_smir::alloc` handles all direct interactions with rustc queries and performs the actual memory allocations, while the `stable_mir::alloc` is responsible for constructing stable components.

makai410 committed:
refactor: rewrite the whole `rustc_smir::context`

This commit removes the `Tables` field from `SmirCtxt`, since borrows of `tables` should only be managed by `SmirInterface`.

This change prevents `SmirCtxt` from holding separate borrows and requires passing `tables` explicitly when needed.

We use the `traits.rs` file to define traits that are used for encapsulating the associated functions in the rustc's internals. This is much easier to use and maintain than directly cramming everything into `SmirCtxt`.

makai410 committed:
refactor: move `convert` module to `stable_mir`

note that this commit delete `convert/error.rs`, we would use `SmirError::from_internal` instead.

**Unresolved questions:**
- There are still a few direct calls to rustc's internals scattered across `impl Stable`s, but most of them appear to be relatively stable, e.g., `mir::interpret::ConstAllocation::inner(self)` and `mir::syntax::SwitchTargets::otherwise(self)`.

makai410 committed:
fix: rewrite some parts of `rustc_smir::builder.rs`

makai410 commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -152,124 +149,3 @@ where
-    type T = Result<T::T, E::T>;
-
-    fn stable(&self, tables: &mut Tables<'_>) -> Self::T {
-        match self {
-            Ok(val) => Ok(val.stable(tables)),
-            Err(error) => Err(error.stable(tables)),
-        }
-    }
-}
-
-impl<'tcx, T> Stable<'tcx> for &[T]
-where
-    T: Stable<'tcx>,
-{
-    type T = Vec<T::T>;
-    fn stable(&self, tables: &mut Tables<'_>) -> Self::T {
-        self.iter().map(|e| e.stable(tables)).collect()
-    }
-}
-
-impl<'tcx, T, U> Stable<'tcx> for (T, U)
-where
-    T: Stable<'tcx>,
-    U: Stable<'tcx>,
-{
-    type T = (T::T, U::T);
-    fn stable(&self, tables: &mut Tables<'_>) -> Self::T {
-        (self.0.stable(tables), self.1.stable(tables))
-    }
-}
-
-impl<'tcx, T> Stable<'tcx> for RangeInclusive<T>
-where
-    T: Stable<'tcx>,
-{
-    type T = RangeInclusive<T::T>;
-    fn stable(&self, tables: &mut Tables<'_>) -> Self::T {
-        RangeInclusive::new(self.start().stable(tables), self.end().stable(tables))
-    }
-}
```
Oops! I accidentally included these changes in this commit, they should have been part of the 'move `convert` module to `stable_mir `' commit instead. Sorry for the confusion🥲

makai410 commented on compiler/rustc_smir/src/rustc_smir/context/mod.rs:
```diff
@@ -0,0 +1,79 @@
+        &self,
+        err: ty::layout::FnAbiError<'tcx>,
+        _span: rustc_span::Span,
+        fn_abi_request: ty::layout::FnAbiRequest<'tcx>,
+    ) -> B::Error {
+        B::Error::new(format!("Failed to get ABI for `{fn_abi_request:?}`: {err:?}"))
+    }
+}
+
+impl<'tcx, B: Bridge> LayoutOfHelpers<'tcx> for Tables<'tcx, B> {
+    type LayoutOfResult = Result<ty::layout::TyAndLayout<'tcx>, B::Error>;
+
+    #[inline]
+    fn handle_layout_err(
+        &self,
+        err: ty::layout::LayoutError<'tcx>,
+        _span: rustc_span::Span,
+        ty: Ty<'tcx>,
+    ) -> B::Error {
+        B::Error::new(format!("Failed to get layout for `{ty}`: {err}"))
+    }
+}
+
+impl<'tcx, B: Bridge> HasTypingEnv<'tcx> for Tables<'tcx, B> {
+    fn typing_env(&self) -> ty::TypingEnv<'tcx> {
+        ty::TypingEnv::fully_monomorphized()
+    }
+}
+
+impl<'tcx, B: Bridge> HasTyCtxt<'tcx> for Tables<'tcx, B> {
+    fn tcx(&self) -> TyCtxt<'tcx> {
+        self.tcx
+    }
+}
+
+impl<'tcx, B: Bridge> HasDataLayout for Tables<'tcx, B> {
+    fn data_layout(&self) -> &rustc_abi::TargetDataLayout {
+        self.tcx.data_layout()
+    }
+}
```
I guess we should implement these traits for `SmirCtxt` instead of `Tables`?

makai410 commented on compiler/rustc_smir/src/rustc_smir/mod.rs:
```diff
@@ -28,28 +28,50 @@ mod builder;
-impl<'tcx> Tables<'tcx> {
-    pub(crate) fn intern_ty(&mut self, ty: Ty<'tcx>) -> stable_mir::ty::Ty {
+pub struct Tables<'tcx, B: Bridge> {
+    tcx: TyCtxt<'tcx>,
```
I think I should remove this `tcx`.

makai410 committed:
refactor: remove the `tcx` field in `Tables`

the only functionality of `Tables` is caching results. this commit moves calls to rustc queries from `Tables` to `SmirCtxt`.

celinval commented:
Sweet! I'll start the review tomorrow. Thanks

makai410 commented on compiler/rustc_smir/src/stable_mir/compiler_interface.rs:
```diff
@@ -5,133 +5,586 @@
-impl<'tcx> SmirInterface<'tcx> {
-    pub(crate) fn entry_fn(&self) -> Option<CrateItem> {
-        self.cx.entry_fn()
+impl<'tcx> Tables<'tcx, BridgeTys> {
```
I was considering moving this `impl Tables` to `rustc_smir`, where we'd define associated types for `***Def` stuffs and traits with associated constructor functions.

makai410 committed:
refactor: unify `Tables` implementation with bridge types and re-export `IndexedVal`

define bridge types for `***Def`s.
consolidate scattered `Tables` implementations into single inherent impl.

makai410 committed:
refactor: move `IndexMap` to `rustc_smir`

we should no longer keep `IndexMap` in `rustc_internal`, as we've decided to migrate `rustc_internal` to `stable_mir` under a feature.

celinval commented on compiler/rustc_smir/src/stable_mir/mod.rs:
```diff
@@ -27,17 +27,21 @@
 use std::fmt::Debug;
 use std::{fmt, io};
 
+use rustc_hir::def::DefKind;
+use rustc_smir::IndexedVal;
```
We could create a stable version of IndexedVal as well to avoid spreading this everywhere. We could just add an  internal trait impl for all types that implement the stable one.

makai410 replied:
I wonder if we can just re-export `IndexedVal` in `stable_mir`^^

makai410 replied:
I wonder if we can just add a
```rust
pub(crate) use rustc_smir::IndexedVal;
```
🤔

celinval replied:
Yeah, that would do it for now

celinval reviewed (commented):
I really like the new name schema. One high level comment, I would prefer keeping StableMIR definitions and logic separate from any sort of conversion and usage of internal rustc code. We could restrict the usage of internal items to be inside specific modules, maybe `rustc_internal` and `convert` modules. What do you think?

makai410 commented:
> I really like the new name schema. One high level comment, I would prefer keeping StableMIR definitions and logic separate from any sort of conversion and usage of internal rustc code. We could restrict the usage of internal items to be inside specific modules, maybe `rustc_internal` and `convert` modules. What do you think?

Ahh I agree. The only problem is `stable_mir/alloc.rs`, which uses some internal items, but I think we can just move all the logic to the `rustc_smir` side^^

makai410 commented on compiler/rustc_smir/src/stable_mir/mod.rs:
```diff
@@ -218,6 +222,43 @@ pub fn all_trait_impls() -> ImplTraitDecls {
     with(|cx| cx.all_trait_impls())
 }
 
+pub(crate) fn new_item_kind(kind: DefKind) -> ItemKind {
```
Yes

celinval commented:
Which internal items are we using in `alloc.rs`? It's probably a mistake anyway, and it should be in `rustc_smir` instead.

celinval commented on compiler/rustc_smir/src/stable_mir/convert/mod.rs:
Can you please also add documentation to this module. I think it's worth mentioning the following:
1. The obvious, that this module holds the logic to convert rustc internal ADTs into stable mir ADTs.
2. The not so obvious, that conversion from stable to internal is not meant to be complete, and it should be added as when needed to be passed as input to `rustc_smir` functions.
3. Some guidelines on how to write conversion code:
    - Avoid calling rustc's internal functions and queries. These should be done via `rustc_smir` APIs.
    - It's OK to access ADT fields directly.

celinval commented on compiler/rustc_smir/src/stable_mir/convert/mod.rs:
```diff
@@ -0,0 +1,126 @@
+    }
+}
+
+/// Trait used to convert between an internal MIR type to a Stable MIR type.
+pub trait Stable<'tcx> {
+    /// The stable representation of the type implementing Stable.
+    type T;
+    /// Converts an object to the equivalent Stable MIR representation.
+    fn stable(
+        &self,
+        tables: &mut Tables<'tcx, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
+    ) -> Self::T;
+}
+
+/// Trait used to translate a stable construct to its rustc counterpart.
+///
+/// This is basically a mirror of [Stable].
+pub trait RustcInternal {
+    type T<'tcx>;
+    fn internal<'tcx>(
+        &self,
+        tables: &mut Tables<'_, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
+    ) -> Self::T<'tcx>;
+}
```
nit: can you move the public declarations to the top of the module?

celinval commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -54,139 +43,11 @@ pub fn stable<'tcx, S: Stable<'tcx>>(item: S) -> S::T {
 ///
 /// This function will panic if StableMIR has not been properly initialized.
-pub fn internal<'tcx, S>(tcx: TyCtxt<'tcx>, item: S) -> S::T<'tcx>
+pub fn internal<'tcx, S>(item: S) -> S::T<'tcx>
```
Actually, this isn't safe. This allows users to create objects with any lifetime `'tcx`, including `'static`. See [this commit](https://github.com/rust-lang/rust/pull/120128/commits/9aace6723572438a94378451793ca37deb768e72) and [this PR](https://github.com/rust-lang/rust/pull/120128) for more context.

makai410 replied:
```suggestion
pub fn internal<'tcx, S>(tcx: TyCtxt<'tcx>, item: S) -> S::T<'tcx>
```

celinval commented on compiler/rustc_smir/src/stable_mir/compiler_interface.rs:
```diff
@@ -5,133 +5,520 @@
-        self.cx.entry_fn()
-    }
-
+pub(crate) trait SmirInterface {
```
Can we get rid of this trait?

makai410 replied:
No. We can't do inherent impl out of `rustc_smir` crate.

celinval replied:
but why do we need inherent impl?

makai410 replied:
umm iiuc, if we get rid of this trait, then we need to do `impl SmirContainer<'tcx, BridgeTys>` in `stable_mir`.

celinval replied:
Ok. We can revisit this later. Thanks

celinval commented on compiler/rustc_smir/src/stable_mir/convert/mod.rs:
```diff
@@ -0,0 +1,126 @@
+    }
+}
+
+/// Trait used to convert between an internal MIR type to a Stable MIR type.
+pub trait Stable<'tcx> {
```
hummm... maybe this module should also go inside `stable_mir::internal` or something like `stable_mir::unstable`. We don't provide any stability guarantees on the `Stable` and `Internal` traits.

```suggestion
/// Trait used to convert between an internal MIR type to a Stable MIR type.
///
/// This trait is currently exposed to users so they can have interoperability between internal MIR
/// and StableMIR constructs. However, they should be used seldomly and they have no influence
/// in this crate semver.
#[doc(hidden)]
pub trait Stable<'tcx> {
```

makai410 replied:
Then... how about we bundle all unstable items that have no stability guarantees into a certain module like `stable_mir::unstable`, including `rustc_internal`?

celinval commented on compiler/rustc_smir/src/stable_mir/convert/mod.rs:
```diff
@@ -0,0 +1,126 @@
+    ) -> Self::T;
+}
+
+/// Trait used to translate a stable construct to its rustc counterpart.
+///
+/// This is basically a mirror of [Stable].
+pub trait RustcInternal {
```
```suggestion
/// Trait used to translate a stable construct to its rustc counterpart.
///
/// This is basically a mirror of [Stable].
///
/// This trait is currently exposed to users so they can have interoperability between internal MIR
/// and StableMIR constructs. They should be used seldomly as they have stability guarantees.
#[doc(hidden)]
pub trait RustcInternal {
```

celinval commented on compiler/rustc_smir/src/stable_mir/convert/mod.rs:
```diff
@@ -0,0 +1,126 @@
+    fn internal<'tcx>(
+        &self,
+        tables: &mut Tables<'_, BridgeTys>,
+        cx: &SmirCtxt<'tcx, BridgeTys>,
```
this should actually be `TyCtxt<'tcx>`

makai410 replied:
Did that mean we allow it to call rustc queries?

celinval replied:
good point... what if we create a trait `TyLift` that invokes `TyCtxt::lift()` method, and pass `impl TyLIft` here.

makai410 replied:
I think it makes sense. btw I guess `RustcInternal` won't be around for long right? So I think allowing it to call rustc queries is also okay since we don't provide any stability guarantees on it anyway.

celinval replied:
I think it will be there. It's not meant for public consumption though

celinval reviewed (commented):
Some more comments. Let me know when you get to address them! Thanks

@rustbot author

makai410 commented:
> One high level comment, I would prefer keeping StableMIR definitions and logic separate from any sort of conversion and usage of internal rustc code.
iiuc, we shouldn't retrieve any internal items even in `compiler_interface.rs`?

makai410 commented:
> One high level comment, I would prefer keeping StableMIR definitions and logic separate from any sort of conversion and usage of internal rustc code. We could restrict the usage of internal items to be inside specific modules, maybe `rustc_internal` and `convert` modules.

iiuc, we shouldn't use any internal items even like `let did = tables[def_id]`? Did that say, we should extract the conversion logic from `compiler_interface.rs`, then move it to somewhere like the `convert` module?

celinval commented:
I can live with `tables[def_id]` since that is part of `rustc_smir` public interface. I think we should minimize manipulation of internal items inside `stable_mir`, so it won't get inadvertently broken.

I think we should avoid calling internal methods in `stable_mir`, e.g.: `Const::new()`, and restrict where we inspect the internals of a type.

bors commented:
:umbrella: The latest upstream changes (presumably #141605) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

makai410 reviewed (commented):
What if we wrap `TyCtxt<'tcx>` in a new `SmirCtxt`? I'm not sure if this approach is correct or not.

celinval commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -197,49 +58,39 @@ pub fn crate_num(item: &stable_mir::Crate) -> CrateNum {
-        let context = ptr as *const SmirCtxt<'_>;
-        let mut tables = unsafe { (*context).0.borrow_mut() };
-        f(&mut *tables)
+        let container = ptr as *const SmirContainer<'tcx, B>;
+        let mut tables = unsafe { (*container).tables.borrow_mut() };
+        let cx = unsafe { (*container).cx.borrow() };
+        f(&mut *tables, &*cx)
```
I'm not sure I follow what you are trying to do here. In the case where `matches!(tcx, None)`, wouldn't `SmirCtxt` be empty?

celinval commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -54,139 +43,11 @@ pub fn stable<'tcx, S: Stable<'tcx>>(item: S) -> S::T {
-    pub(crate) fn layout_id(&mut self, layout: rustc_abi::Layout<'tcx>) -> Layout {
-        self.layouts.create_or_fetch(layout)
-    }
+    with_container(|tables, cx| item.internal(tables, cx))
```
The idea is for `internal` calls, we want the `tcx` that the user provided.

```suggestion
    with_container(|tables, _| item.internal(tables, SmirCtxt::new(tcx)))
```
If we implement the `TyLift` trait, this would become:
```suggestion
    with_container(|tables, _| item.internal(tables, tcx))
```

makai410 replied:
My concern is that `tcx.lift()` is not the only method being used in `impl RustcInternal`s. There are other calls to rustc queries e.g. `tcx.mk_args_from_iter()`, `tcx.mk_pat()`, `tcx.mk_poly_existential_predicates()`, which are not defined in the `ty::Lift` trait but live in inherent impl of `TyCtxt`. I was wondering should we allow these calls? If not, then the approach of passing `impl TyLift` would fail.

IIUC, the unsoundness comes from the lifetime. To solve this issue requires passing a `TyCtxt<'tcx>` to `rustc_internal::internal()`. Wrapping the provided `TyCtxt<'tcx>` in a new `SmirCtxt` keeps the lifetime, which would also make `RustcInternal` insulated from direct calls to rustc queries.

celinval replied:
By `TyLift`, I mean that's a private trait that we define inside `stable_mir::internal` that allow us to control which methods are OK to call from `internal` methods. `mk_args_from_iter()`, `mk_pat()` and other `mk_*` functions like `lift` are about ensuring things are in the arena for that TyCtxt, so we should include those. Maybe `TyLift` is not the best name. We could also name it after `Arena`

makai410 replied:
Ah then it makes sense. In that case `TyLift` would be only for `RustcInternal`, any other attempts to rustc queries should be proxied to `SmirCtxt`.

makai410 replied:
What if we call this trait `InternalCx`?

makai410 committed:
fix: resolve the unsoundness

add a new trait `InternalCx`, which defines the methods that are fine to call from `RustcInternal`. `RustcInternal::internal()` then takes a `impl InternalCx<'tcx>` instead of `TyCtxt<'tcx>`.

make `tcx` in `SmirCtxt` public, since we need to pass it to `RustcInternal::internal()` in `SmirInterface`.

makai410 committed:
refactor: move `convert` to `unstable`

We want to keep StableMIR definitions and logic separate from any sort of conversion and usage of internal rustc code. So we bundle all unstable items that have no stability guarantees into `stable_mir::unstable`.

makai410 committed:
refactor: avoid calling internal functions in `predicates_of()` and `explicit_predicates_of()`

makai410 commented on compiler/rustc_smir/src/rustc_smir/context/impls.rs:
```diff
@@ -197,12 +197,32 @@ impl<'tcx, B: Bridge> SmirCtxt<'tcx, B> {
+                .iter()
+                .map(|(clause, span)| (clause.as_predicate().kind().skip_binder(), *span))
+                .collect(),
+        )
```
Pretty unsure if this a good approach but I can't figure out a perfect way🤷‍♂️

makai410 committed:
refactor: don't use rustc_type_ir directly

makai410 committed:
refactor: rewrite `adt_repr()`, `adt_discr_for_variant()` and `coroutine_discr_for_variant()`

celinval commented on compiler/rustc_smir/src/rustc_smir/alloc.rs:
```diff
@@ -1,102 +1,79 @@
-use stable_mir::Error;
-use stable_mir::mir::Mutability;
-use stable_mir::ty::{Allocation, ProvenanceMap};
+//! Internal memory allocator implementation for StableMIR.
```
nice!

celinval commented on compiler/rustc_smir/src/rustc_smir/alloc.rs:
```diff
@@ -1,102 +1,79 @@
     alloc_range: AllocRange,
-    tables: &mut Tables<'tcx>,
-) -> Allocation {
+) -> (Vec<Option<u8>>, Vec<(usize, AllocId)>) {
```
would it make sense to create a `rustc_smir` type here to encapsulate this?

celinval commented on compiler/rustc_smir/src/rustc_internal/mod.rs:
```diff
@@ -4,28 +4,17 @@
-use stable_mir::abi::Layout;
-use stable_mir::compiler_interface::SmirInterface;
-use stable_mir::ty::IndexedVal;
+use stable_mir::unstable::{RustcInternal, Stable};
```
It seems that rustc_smir is still depending on stable_mir here. I think we need to move `stable` and `internal` functions to `stable_mir::unstable`, right? Or are you planning to define the traits here?

BTW, I'm OK with leaving this to a follow up PR too.

celinval reviewed (commented):
I finally got to do another pass. Just a few comments and I think we should be good.

bors commented:
:umbrella: The latest upstream changes (presumably #142997) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

makai410 committed:
refactor: implement a new bridge trait `Allocation`

bors commented:
:umbrella: The latest upstream changes (presumably #143026) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #142974) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #143363) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
